[SLP]Add runtime stride support for strided loads.

alexey-bataev opened the pull request:
Added support for runtime strides.

alexey-bataev committed:
[𝘀𝗽𝗿] initial version

Created using spr 1.3.5

alexey-bataev committed:
Rebase

Created using spr 1.3.5

alexey-bataev reviewed (commented):
Ping!

alexey-bataev committed:
Rebase

Created using spr 1.3.5

preames commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -3890,6 +3891,122 @@ static bool isReverseOrder(ArrayRef<unsigned> Order) {
+    const SCEV *Diff1 = SE.getMinusSCEV(PtrSCEVB, PtrSCEV);
+    if (!Diff1 || isa<SCEVCouldNotCompute>(Diff1))
+      return std::nullopt;
+    if (Diff1->isNonConstantNegative()) {
```
You can swap the order of the subtract and use isKnownPositive here.

alexey-bataev replied:
I need to check that it is also not a constant

preames commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -3890,6 +3891,122 @@ static bool isReverseOrder(ArrayRef<unsigned> Order) {
+      PtrSCEVB = PtrSCEV;
+      continue;
+    }
+  }
```
Ok, there's something significant I'm missing here.

PtrSCEVA and PtrSCEVB are going to be (after the loop above) our best guess as to the bounds on the access pattern, but since we may have failed to be able to compare some of the pointers, they are not a true set of bounds.  

How does using those guessed bounds to compute a guessed stride guaranteed to be correct?

alexey-bataev replied:
If at least single pointer cannot be proved to be a "strided" pointer, the code just early exits the function, returning std::nulopt (i.e. we cannot consider the pointers top be the source of strided accesses)

preames reviewed (commented):
Inline comments - sorry, these had actually been made a bit ago, but I apparently forgot to submit the review so they were still pending.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -3890,6 +3891,122 @@ static bool isReverseOrder(ArrayRef<unsigned> Order) {
+      continue;
+    }
+    const SCEV *Diff = SE.getMinusSCEV(PtrSCEV, PtrSCEVA);
+    if (!Diff || isa<SCEVCouldNotCompute>(Diff))
```
Yep, you're right, forgot to remove

alexey-bataev committed:
Rebase, address comments

Created using spr 1.3.5

alexey-bataev committed:
Rebase

Created using spr 1.3.5

alexey-bataev committed:
Rebase

Created using spr 1.3.5

preames reviewed (approved):
LGTM

I'll comment that I find the guess and check style logic used in this patch to be very confusing without clear comments as to exactly what's going on.  I think you could make this much clearer if you used appropriate variable names, or at least some clear comments about the point where you transfer from best effort guessing to checking your result and relying on it's correctness.

alexey-bataev commented:
> LGTM
> 
> I'll comment that I find the guess and check style logic used in this patch to be very confusing without clear comments as to exactly what's going on. I think you could make this much clearer if you used appropriate variable names, or at least some clear comments about the point where you transfer from best effort guessing to checking your result and relying on it's correctness.

Renamed the variables, added some more comments

alexey-bataev committed:
Rebase, address comments

Created using spr 1.3.5
